### PR TITLE
CA-220505: Destroy stale VM record on destination pool on failed …

### PR DESCRIPTION
**Pull request by sharady**
…SXM.

Cancelling cross-pool SXM ends up in destroying VDIs and stale VM record
on destination Pool.
After cancellation, If VM exists on source Pool then destroy the stale VM
record on destination Pool.

Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 1]
```diff
@@ -966,6 +966,22 @@ let migrate_send'  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
       debug "Mirror failed for VDI: %s" failed_vdi;
       raise (Api_errors.Server_error(Api_errors.mirror_failed,[Ref.string_of vconf.vdi]))
     end;
+
+    if (TaskHelper.is_cancelling ~__context) && not is_intra_pool then
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: You are using parentheses for the `TaskHelper.is_canceling` function application but not for `not`. You don't need parentheses in either case (it's fine to write `if f x && not y then .. else ..`) but I think this should be consistent.

File: ocaml/xapi/xapi_vm_migrate.ml [hunk 2]
```diff
@@ -966,6 +966,22 @@ let migrate_send'  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
...
+      begin match e with
+      | Storage_interface.Does_not_exist(_) ->
+        if (Db.is_valid_ref __context vm) then begin
```
Comment (lindig) [inline ocaml/xapi/xapi_vm_migrate.ml]: You are using parentheses again here. They are not strictly needed and the convention is not to use them for simple if-expressions.

**Comment by jonludlam** [comment]
We're matching some very specific conditions here - it seems odd that we only destroy the remote VM if we catch 'Storage_interface.Does_not_exist' - can you justify that?

File: ocaml/xapi/xapi_vm_migrate.ml (hunk 2 above)
Comment (sharady) [inline ocaml/xapi/xapi_vm_migrate.ml]: @lindig Thanks for the review, I have removed the parentheses.

**Comment by sharady** [comment]
@jonludlam I have seen this specific error `Storage_interface.Does_not_exist` in case of CA-220505. Where cancellation happens after VM metadata import is complete and VDI mirror is done. I have posted the trace for it on the ticket. Might need to discuss it further if any other cases are missing which i am also not sure. Thanks.

**Comment by sharady** [comment]
@jonludlam  Thanks for the suggestion, I have updated the PR accordingly. Please have a look :)

**Comment by lindig** [comment]
Paging @jonludlam ;)

**Comment by jonludlam** [comment]
Looks good now, thanks!

**Pull request merged**